Kitodo/src/main/java/org/kitodo/production/services/file/FileService.java

```diff
@@ -1183,39 +1200,92 @@ private void addNewURIsToExistingPhysicalDivisions(Map<String, Map<Subfolder, UR
         }
     }
 
+    private void removeMissingMediaFromWorkpiece(List<String> mediaToRemove, Workpiece workpiece,
+                                                 Collection<Subfolder> subfolders) {
+        List<PhysicalDivision> pages = workpiece.getAllPhysicalDivisionChildrenFilteredByTypePageAndSorted();
+        for (String removal : mediaToRemove.stream().filter(Objects::nonNull).collect(Collectors.toList())) {
+            for (PhysicalDivision page : pages) {
+                if (removal.equals(getCanonical(subfolders, page))) {
+                    workpiece.getPhysicalStructure().getChildren().remove(page);
+                    for (LogicalDivision structuralElement : page.getLogicalDivisions()) {
+                        structuralElement.getViews().removeIf(view -> view.getPhysicalDivision().equals(page));
+                    }
+                    page.getLogicalDivisions().clear();
+                    LinkedList<PhysicalDivision> ancestors = MetadataEditor
+                            .getAncestorsOfPhysicalDivision(page, workpiece.getPhysicalStructure());
+                    if (!ancestors.isEmpty()) {
+                        PhysicalDivision parent = ancestors.getLast();
+                        parent.getChildren().remove(page);
+                    }
+                }
+            }
+        }
+        if (!mediaToRemove.isEmpty()) {
+            int i = 1;
+            for (PhysicalDivision division : workpiece.getAllPhysicalDivisionChildrenFilteredByTypePageAndSorted()) {
+                division.setOrder(i);
+                i++;
+            }
+        }
+    }
+
+    private String getCanonical(Collection<Subfolder> folders, PhysicalDivision division) {
+        Map<MediaVariant, URI> mediaFiles = division.getMediaFiles();
+        for (Subfolder folder : folders) {
+            for (URI mediaUri : mediaFiles.values()) {
+                if (mediaUri.getPath().startsWith(folder.getFolder().getRelativePath())) {
+                    String canonical = folder.getCanonical(mediaUri);
+                    if (Objects.nonNull(canonical)) {
+                        return canonical;
+                    }
+                }
+            }
+        }
+        return null;
```

Comment:
According to the coding rules `null` return should be avoided. If I see the current usage right than an empty string (`""`) should be good enough too or?

Reply:
Thanks, changed!

Reply:
You added a check with the old stringutils class. Can you switch to stringutlils3 class instead or we need to add this old library as dependency again.

Reply:
> You added a check with the old stringutils class. Can you switch to stringutils3 class instead or we need to add this old library as dependency again.

:dizzy_face: 

thanks for pointing this out 😅